client/consensus/common/src/level_monitor.rs

```diff
@@ -0,0 +1,247 @@
+	/// the limit passed to the constructor.
+	///
+	/// If the given level is found to have a number of blocks greater than or equal the limit
+	/// then the limit is enforced by chosing one (or more) blocks to remove.
+	/// The removal strategy is driven by the block freshess.
+	///
+	/// A block freshness is determined by the most recent leaf freshness descending from the block
+	/// itself. In other words its freshness is equal to its more "fresh" descendant.
+	///
+	/// The least "fresh" blocks are eventually removed.
+	pub fn enforce_limit(&mut self, number: NumberFor<Block>) {
+		if self.import_counter == 0 {
+			// First call detected, restore using the backend persistent information.
+			self.restore();
+		}
+
+		let level = self.levels.entry(number).or_default();
+		if level.len() < self.level_limit {
+			return
+		}
+
+		log::debug!(
+			target: "parachain",
+			"Detected leaves overflow at height {}, removing obsolete blocks", number
+		);
+
+		let mut leaves = self.backend.blockchain().leaves().unwrap_or_default();
+		// Sort leaves by freshness (most recent first).
+		leaves.sort_unstable_by(|a, b| self.freshness.get(a).cmp(&self.freshness.get(b)));
+
+		let remove_count = level.len() - self.level_limit + 1;
+
+		// This may not be the most efficient way to remove **multiple** entries, but is the clear
+		// one. Should be considered that in normal conditions the number of blocks to remove is 0
+		// or 1, it is not worth to complicate the code too much.
+		(0..remove_count).for_each(|_| self.remove_block(number, &leaves));
+	}
+
+	/// This will effectively search for the less fresh block and removes it along with all
+	/// its descendants.
+	fn remove_block(&mut self, number: NumberFor<Block>, leaves: &[Block::Hash]) {
+		let mut candidate_leaves_routes = vec![];
+		let mut candidate_fresher_leaf_idx = usize::MAX;
+
+		let mut leaves_routes = vec![];
+
+		let blockchain = self.backend.blockchain();
+		let best_hash = blockchain.info().best_hash;
+
+		for blk_hash in self.levels.entry(number).or_default().iter() {
+			let mut fresher_leaf_idx = usize::MAX;
+			// Start checking the most fresh leaf (the last)
+			// Here leaf index is synonym of its freshness, that is the greater the index the
+			// fresher it is.
+			for (leaf_idx, leaf_hash) in leaves.iter().enumerate().rev() {
+				let route = match sp_blockchain::tree_route(blockchain, *blk_hash, *leaf_hash) {
+					Ok(route) => route,
+					Err(e) => {
+						log::warn!(
+							target: "parachain",
+							"Unable getting route from {} to {}: {}",
+							blk_hash, leaf_hash, e,
+						);
+						continue
+					},
+				};
+				if route.retracted().is_empty() {
+					if leaf_idx >= candidate_fresher_leaf_idx ||
+						route.common_block().hash == best_hash ||
+						route.enacted().iter().any(|entry| entry.hash == best_hash)
+					{
+						leaves_routes.clear();
+						break
+					}
+
+					if leaves_routes.is_empty() {
+						fresher_leaf_idx = leaf_idx;
```

Comment:
Not entirely sure, but don't we also need to update `fresher_leaf_idx` when we come across `leaf_idx > fresher_leaf_idx`?

Reply:
Code has been just updated to be more readable and optimized. Let me know if it is clear now